src/flint/types/fq_default.pyx

```diff
@@ -0,0 +1,886 @@
+
+        if check_prime and not prime.is_prime():
+            raise ValueError("characteristic is not prime")
+
+        # the degree must be strictly positive
+        if d < 1:
+            raise ValueError(f"the degree must be positive, got {d = }")
+
+        # Cython type conversion and context initalisation
+        self._c_set_from_order(prime, d, var, fq_type)
+
+    cdef _c_set_from_modulus(self, modulus, char *var,
+                             fq_default_type fq_type=fq_default_type.DEFAULT):
+        self.var = var
+        if typecheck(modulus, fmpz_mod_poly):
+            fq_default_ctx_init_modulus_type(self.val, (<fmpz_mod_poly>modulus).val,
+                                             (<fmpz_mod_poly>modulus).ctx.mod.val, self.var, fq_type)
+        elif typecheck(modulus, nmod_poly):
+            fq_default_ctx_init_modulus_nmod_type(self.val, (<nmod_poly>modulus).val,
+                                                  self.var, fq_type)
+        else:
+            raise TypeError(f"modulus must be fmpz_mod_poly or nmod_poly, got {modulus!r}")
+        self._initialized = True
+
+    def set_from_modulus(self, modulus, var,
+                         fq_type=fq_default_type.DEFAULT, check_modulus=True):
+        """
+        Construct a context for a finite field from an irreducible polynomial.
+
+        `modulus` may be of type :class:`~.fmpz_mod_poly` or :class:`~.nmod_poly`.
+
+        `var` is a name for the ring generator of this field over the prime field.
+
+        The optional parameter `type` select the implementation. The
+        possible values are:
+
+        - `fq_default_ctx.DEFAULT`: implementation automatically decided by Flint (default),
+        - `fq_default_ctx.FQ_ZECH`: Use `fq_zech_t`,
+        - `fq_default_ctx.FQ_NMOD`: Use `fq_nmod_t`,
+        - `fq_default_ctx.FQ`: Use `fq_t`.
+        """
+        if check_modulus and not modulus.is_irreducible():
+            raise ValueError("modulus must be irreducible")
+
+        # Cython type conversion and context initalisation
+        self._c_set_from_modulus(modulus, var, fq_type)
+
+    @property
+    def fq_type(self):
+        """
+        Return the implementation of this context. It is one of:
+
+        - 1: `fq_default_ctx.FQ_ZECH`: Using `fq_zech_t`,
+        - 2: `fq_default_ctx.FQ_NMOD`: Using `fq_nmod_t`,
+        - 3: `fq_default_ctx.FQ`: Using `fq_t`.
+        """
+        return fq_default_ctx_type(self.val)
+
+    @property
+    def fq_type_str(self):
+        """
+        Return the string implementation of this context. It is one of:
+
+        - `fq_default_ctx.FQ_ZECH`: Using `fq_zech_t`,
+        - `fq_default_ctx.FQ_NMOD`: Using `fq_nmod_t`,
+        - `fq_default_ctx.FQ`: Using `fq_t`.
+        """
+        FQ_TYPES = {
+            1 : "FQ_ZECH",
+            2 : "FQ_NMOD",
+            3 : "FQ",
+            4 : "NMOD",
+            5 : "FMPZ_MOD",
+        }
+        return FQ_TYPES[self.fq_type]
```

Comment:
I think it should be possible to get the string from the enum if `self.fq_type` returns the enum like:
```
return self.fq_type._name_
```
Then we don't need to duplicate the names and numbers here.